### PR TITLE
Added term document frequency map to persistent settings

### LINKED ISSUE #116: Implement storage for term document frequencies
Keep a dictionary that maps each word user has read to the number of documents it has occurred in. We can then use this for TF-IDF to better determine keywords.

### PR DESCRIPTION
**Pull request by benfrancis314**
This is a necessary component for the transition to TF-IDF for keywords. For each word the user encounters, it keeps track of how many documents it has appeared in. If you want to see what this look like, add console.log(termDocumentFreq) under setTermDocumentFreq.

Notes: 
- While the "DF" in TF-IDF stands for "document frequency", I called it "term document frequency" because I didn't think document frequency was intuitively understandable enough. Document frequency of what? Term document frequency references for each term the frequency of documents it has appeared in. 
- I am storing it locally for now, then we can transition it to .sync long term storage later. I'm still going to associate it with the issue #116, bc the gist of it is here, I just want to have the new TF-IDF keywords all fleshed out before switching to .sync. 

Close #116

### CODE REVIEW

**Review by sjonany — CHANGES_REQUESTED** [review]
This idea is interesting, but I think it's worth discussing first before proceeding w/ the review.

1. Is this the desired behavior?
- User won't get creeped out
- Maybe users want keywords based off only the currently article

2. Unlike the other settings, this setting can grow in size.
Either set an explicit bound on the size, or convince the code readers that users won't have to worry about hitting storage limits.
-- But if you have a limit, then might as well enforce the bound programmatically anyway.

3. On the point of feature creep, my request is that before we introduce a new feature, we go through clickups and make sure that a new feature is what we want to add rn.

**Comment by benfrancis314** [comment]
**1**. [Part A] Users won't know about this, [Part B] This will still base the keywords off the current article, the accumulation of other documents just helps determine what common English words are (like a much more sophisticated version of excluding stop words, this inversely weights common words). Enough docs should remove any initial bias of the DF from certain topics. If we pursue, we'd probably want to initialize with a baseline dict as well. 

**2**. True; two important questions on storage limit: I. )  Is there a local/sync storage limit? (Different from bandwidth limit on each set/get) [Answer: Yes, approx 1 kB, see: https://developer.chrome.com/extensions/storage#property-sync] What is estimate of how much space such a dictionary would take up, in a worst case scenario? (Answer: ~400,000 words in English dictionary, avg adult vocab is 20-40,000, so probably would be a problem). Alternative method could be hardcoding a static baseline DF in, if we do want to pursue the TF-IDF strategy, which does seem to be the most common strategy. Also could look at existing libraries, but I like the idea of doing this in-house bc it doesn't seem intrinsically complicated and I think having a good control over the process could let us do other cool things later. With that said, this is common so a lib is a very legitimate option

II. ) Right now (using the strategy of storing DF in storage.local or storage.sync) the easiest way is to get the old document frequency duct, update it, then send it back. BUT, a much better way would be to just add new words to the existing dictionary in the settings, instead of getting and sending the entire dict every time. This would be significantly faster and not run into bandwidth problems; just don't know how to do bc not explicitly mentioned in https://developer.chrome.com/extensions/storage, but almost certainly possible. Just wanted to get out first push with POC

**3**. I wouldn't describe this as a new feature, but rather improving the keyword algorithm. So a feature being something that interfaces with the user; this changes nothing from the user perspective, but would just make the keywords more accurate.

File: extension/lib/doc.js [hunk 1]
```diff
@@ -168,10 +172,29 @@ class Doc {
...
+
         this.termFreq = termFreq; // Set class attribute "termFreq"
         return total_words // Set total words for use elsewhere (like Display)
     }
 
+    updateTermDocumentFreq(termFreq, termDocumentFreq, cb) {
+        // For each unique word in doc; add to termDocumentFreq dict or add one to its frequency
+        for (const word in termFreq) {
+            if (word in termDocumentFreq) {
+                termDocumentFreq[word]++
```
Comment (sjonany) [inline extension/lib/doc.js:189]: for both if and else -- don't use want to add the freq, instead of +1?

File: extension/lib/settings_wrapper.js [hunk 2]
```diff
@@ -98,7 +116,7 @@ class Settings {
 			cb(speed);
 		})
 	}
-
+	// See setDefaultTrackerSettigns().
```
Comment (sjonany) [inline extension/lib/settings_wrapper.js:120]: Remove this?

**Review by sjonany — CHANGES_REQUESTED** [review]

File: extension/lib/doc.js (hunk 1 above)
Comment (sjonany) [reply extension/lib/doc.js:189]: Oh sorry this is doc freq. My bad.

**Comment by sjonany** [comment]
[Feature Creep?] That's fair. -- end subthread

[Desired behavior?] It's unclear to me if the current approach is good.
1. Ez tech fix
- Current approach doesn't take into account if the _same_ document is being reread

2. Personalization without a [good baseline profile] for new user means initial keywords are unlikely to be good.
Any stopwords missed in the blacklist will always have high TFIDF, until they have read enough articles.

3. Unclear if we want to maintain these [stopwords, TFIDF] logic ourselves or just delegate it to a library
- (haven't explored) https://github.com/retextjs/retext-keywords

4. Unclear if TF-IDF optimizes for "keywords that will improve skimming"
The "keywords" that we are interested in here are ones that when skimmed, can still retain the most info.
This might correlate w/ the tfidf-detected keywords, but I'm unsure. Heck, even a "subject, verb, object" highlighter could arguably be more readable than if tfidf always skips the common verbs.
Request: Either discuss live or create a document so we can discuss on gdocs. I think the main goal is "keywords such that on skimming, reading just these keywords allow u to retain most info". W/ the given goal, discuss approaches, and pick the best one.


[Storage]
The parsed web text might not be as clean as the english language, especially w/ parsing errors, or if the webpage is non sensical. E.g. a static .txt file of all words in an english dictionary.
Probably won't crash, but please leave a TODO: to actually handle this case explicitly so we won't forget someday if someone's asking why their app doesn't load because they overloaded their tfidf storage.
Request: Leave TODO:

**Comment by sjonany** [comment]
Alternatively, if from briefly trying around, this is >> whatever we have right now, then feel free to check it in after leaving 2 TODOs, then create a new clickup item w/ these discussion points

**Comment by sjonany** [comment]
Ben's point: What if user's profile is stuck in a crappy state? How to recover?

**Comment by benfrancis314** [comment]
Comment on the above: 

- I haven't found anything more _readily_ available to use than just completing this option. An extension such as https://github.com/retextjs/retext-keywords would not be so hard, but it is meant for node. Additionally, there's no direct indication that it would be much better than what we currently have. See: https://retextjs.github.io/retext-keywords/
- With that said, TF-IDF has a pretty strong argument for at least being confidently better than our current system, and it would be very easy to integrate into our current system. The code in this PR does most of the hard part, but its biggest problem appears to be storage. 
- It turns out the local storage limit is ~5 MB, which actually should be good enough for now. So my plan is to finish this option and see where it is deficient, largely since it is close to being done anyway.
- If we have a good enough DF dictionary, it should be pretty hard to screw up (and is either way a corner case)
- Will add TODOs
- I think we discussed in person about the above, right?
- Other benefit: if we handle the TF-IDF in house, it's tweakable. 
-TL;DR -> This is almost done anyway, guaranteed increase of value, storage not as much of a problem as I thought, going to give it a whirl
